lib/torrent.js

```diff
@@ -1599,29 +1599,33 @@ class Torrent extends EventEmitter {
         if (self.destroyed) return
 
         if (hash === self._hashes[index]) {
-          if (!self.pieces[index]) return
           self._debug('piece verified %s', index)
 
-          self.pieces[index] = null
-          self._reservations[index] = null
-          self.bitfield.set(index, true)
-
           self.store.put(index, buf, err => {
-            if (err) self._destroy(err)
-          })
-
-          self.wires.forEach(wire => {
-            wire.have(index)
+            if (err) {
+              // Not sure what else to do here if store threw up.
+              // try to download again?
+              self.pieces[index] = new Piece(piece.length)
+              self.bitfield.set(index, false)
+              self.select(index, index, 1)
+              self.emit('warning', err)
+            } else {
+              self.pieces[index] = null
+              self._markVerified(index)
+              self.wires.forEach(wire => {
+                wire.have(index)
+              })
```

Comment:
@jhiesey The error path is one part, but the other big part of this change is marking the piece as verified when storage is done with the put.